src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -247,13 +254,20 @@ public boolean cancel(final CauseOfInterruption... causes) {
             t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
                 @Override
                 public void onSuccess(CpsThreadGroup g) {
+                    if (g.threads.size() == 1) {
+                        // Early exit for trivial case
+                        g.threads.firstEntry().getValue().stop(stopped);
```

Comment:
Again this is potentially wrong.

(Anyway how often is `cancel` called? Not often at all. So not a priority for little optimizations.)